src/internal_telemetry/allocations/allocator/tracing_allocator.rs

```diff
@@ -48,40 +48,46 @@ impl<A: GlobalAlloc, T: Tracer> GroupedTraceableAllocator<A, T> {
 unsafe impl<A: GlobalAlloc, T: Tracer> GlobalAlloc for GroupedTraceableAllocator<A, T> {
     #[inline]
     unsafe fn alloc(&self, object_layout: Layout) -> *mut u8 {
-        let (group_id_ptr, object_ptr, _wrapped_layout) =
-            self.get_wrapped_allocation(object_layout);
-        let object_size = object_layout.size();
-        // Group id value of zero implies allocations tracking was disabled
-        // during this allocation. We override this if allocations were in fact enabled.
-        group_id_ptr.write(0);
         if TRACK_ALLOCATIONS.load(Ordering::Relaxed) {
+            let (group_id_ptr, object_ptr, _wrapped_layout) =
+                self.get_wrapped_allocation(object_layout);
+            let object_size = object_layout.size();
+            // Group id value of zero implies allocations tracking was disabled
+            // during this allocation. We override this if allocations were in fact enabled.
+            group_id_ptr.write(0);
+            // Don't trace during startup
+            if STARTUP.load(Ordering::Relaxed) {
+                return object_ptr;
+            }
             try_with_suspended_allocation_group(
                 #[inline(always)]
                 |group_id| {
                     group_id_ptr.write(group_id.as_raw());
                     self.tracer.trace_allocation(object_size, group_id);
                 },
             );
+            object_ptr
+        } else {
+            self.allocator.alloc(object_layout)
```

Comment:
I think it would be better to structure this code as:

```
if !tracking_allocations {
  return self.allocator.alloc(...)
}

<do a tracked allocation>
```

Right now, it's a little hard to follow the logic.